File: backend/questions.zig
const std = @import("std");
const http = @import("apple_pie");
const sqlite = @import("sqlite");

const json = std.json;

const Store = @import("sessions/Store.zig");
const Context = @import("Context.zig");
const printTime = @import("time.zig").printTime;

const responses = @import("responses.zig");
const forbidden = responses.forbidden;
const badRequest = responses.badRequest;
const ok = responses.ok;

const max_question_len = 500;

const create_table_questions_query =
    \\CREATE TABLE IF NOT EXISTS "questions" (
    \\  "id"  INTEGER NOT NULL UNIQUE,
    \\  "text"  TEXT NOT NULL,
    \\  "upvotes"  INTEGER NOT NULL,
    \\  "state"  INTEGER NOT NULL,
    \\  "created_at"  INTEGER NOT NULL,
    \\  "modified_at"  INTEGER NOT NULL,
    \\  "answering_at"  INTEGER NOT NULL,
    \\  "answered_at"  INTEGER NOT NULL,
    \\  PRIMARY KEY("id" AUTOINCREMENT)
    \\);
;

const insert_question_query =
    \\INSERT INTO questions(
    \\  "text",
    \\  "upvotes",
    \\  "state",
    \\  "created_at",
    \\  "modified_at",
    \\  "answering_at",
    \\  "answered_at"
    \\) VALUES(
    \\  @text,
    \\  @upvotes,
    \\  @state,
    \\  @created_at,
    \\  @modified_at,
    \\  @answering_at,
    \\  @answered_at
    \\);
;

const list_questions_visible_query = std.fmt.comptimePrint(
    \\SELECT
    \\  "id",
    \\  "text",
    \\  "upvotes",
    \\  "state",
    \\  "created_at",
    \\  "modified_at",
    \\  "answering_at",
    \\  "answered_at"
    \\FROM questions
    \\WHERE
    \\  state = {} OR
    \\  state = {} OR
    \\  state = {}
, .{
    @enumToInt(QuestionState.unanswered),
    @enumToInt(QuestionState.answering),
    @enumToInt(QuestionState.answered),
});

const list_questions_all_query =
    \\SELECT
    \\  "id",
    \\  "text",
    \\  "upvotes",
    \\  "state",
    \\  "created_at",
    \\  "modified_at",
    \\  "answering_at",
    \\  "answered_at"
    \\FROM questions
;

const increment_upvotes_query =
    \\UPDATE questions SET upvotes = upvotes + 1 WHERE id = @id;
;

const modify_question_query =
    \\UPDATE questions SET
    \\  state = @state,
    \\  modified_at = @modified_at
    \\WHERE id = @id;
;

const answering_question_query = std.fmt.comptimePrint(
    \\UPDATE questions SET
    \\  state = {},
    \\  answering_at = @answering_at
    \\WHERE id = @id;
, .{
    @enumToInt(QuestionState.answering),
});

const answered_question_query = std.fmt.comptimePrint(
    \\UPDATE questions SET
    \\  state = {},
    \\  answered_at = @answered_at
    \\WHERE id = @id;
, .{
    @enumToInt(QuestionState.answered),
});

const delete_all_questions_query =
    \\DELETE FROM questions;
;

const QuestionState = enum(u32) {
    hidden,
    unanswered,
    deleted,
    answering,
    answered,

    fn toString(self: QuestionState) []const u8 {
        return switch (self) {
            .hidden => "hidden",
            .unanswered => "unanswered",
            .deleted => "deleted",
            .answering => "answering",
            .answered => "answered",
        };
    }
};

const Question = struct {
    id: u64 = 0, // TODO remove default value
    text: []const u8,
    upvotes: u32 = 0,
    state: QuestionState = .unanswered,
    created_at: i64,
    modified_at: i64,
    answering_at: i64,
    answered_at: i64,
};
// TODO remove
const QuestionInternal = struct {
    text: []const u8,
    upvotes: u32 = 0,
    state: u32 = @enumToInt(QuestionState.unanswered),
    created_at: i64,
    modified_at: i64 = -1,
    answering_at: i64 = -1,
    answered_at: i64 = -1,
};
const QuestionInternalWithoutId = struct {
    text: sqlite.Text,
    upvotes: u32 = 0,
    state: u32 = @enumToInt(QuestionState.unanswered),
    created_at: i64,
    modified_at: i64 = -1,
    answering_at: i64 = -1,
    answered_at: i64 = -1,
};
// TODO rename to `QuestionInternal`
const QuestionInternalSqlite = struct {
    id: u64,
    text: sqlite.Text,
    upvotes: u32 = 0,
    state: u32 = @enumToInt(QuestionState.unanswered),
    created_at: i64,
    modified_at: i64 = -1,
    answering_at: i64 = -1,
    answered_at: i64 = -1,
};

const QuestionIterator = struct {
    allocator: std.mem.Allocator,
    statement: sqlite.DynamicStatement,
    iterator: sqlite.Iterator(QuestionInternalSqlite),

    fn init(allocator: std.mem.Allocator, db: *sqlite.Db, include_hidden: bool) !QuestionIterator {
        const query = if (include_hidden)
            list_questions_all_query
        else
            list_questions_visible_query;

        var statement = try db.prepareDynamic(query);

        return QuestionIterator{
            .allocator = allocator,
            .statement = statement,
            .iterator = try statement.iterator(QuestionInternalSqlite, .{}),
        };
    }

    fn deinit(self: *QuestionIterator) void {
        self.statement.deinit();
    }

    fn next(self: *QuestionIterator) !?Question {
        const question_internal_maybe = try self.iterator.nextAlloc(self.allocator, .{});
        const question_internal = question_internal_maybe orelse return null;

        return Question{
            .id = question_internal.id,
            .text = question_internal.text.data,
            .upvotes = question_internal.upvotes,
            .state = std.meta.intToEnum(QuestionState, question_internal.state) catch .deleted,
            .created_at = question_internal.created_at,
            .modified_at = question_internal.modified_at,
            .answering_at = question_internal.answering_at,
            .answered_at = question_internal.answered_at,
        };
    }
};

pub fn initializeDatabase(db: *sqlite.Db) !void {
    try db.exec(create_table_questions_query, .{}, .{});
}

pub fn listQuestions(ctx: *Context, response: *http.Response, request: http.Request) !void {
    const allocator = request.arena;

    var store = Store{ .redis_client = &ctx.redis_client };
    var session = store.get(allocator, request, "nochfragen_session");
    const logged_in = (try session.get(bool, "authenticated")) orelse false;

    var iter = try QuestionIterator.init(allocator, &ctx.db, logged_in);
    defer iter.deinit();

    var json_write_stream = std.json.writeStream(response.writer(), 4);
    try json_write_stream.beginArray();

    while (try iter.next()) |question| {
        if (question.state == .deleted) continue;

        const str_id = try std.fmt.allocPrint(allocator, "question:{}", .{question.id});
        const upvoted = (try session.get(bool, str_id)) orelse false;

        try json_write_stream.arrayElem();
        try json_write_stream.beginObject();

        try json_write_stream.objectField("id");
        try json_write_stream.emitNumber(question.id);

        try json_write_stream.objectField("text");
        try json_write_stream.emitString(question.text);

        try json_write_stream.objectField("upvotes");
        try json_write_stream.emitNumber(question.upvotes);

        try json_write_stream.objectField("state");
        try json_write_stream.emitNumber(@enumToInt(question.state));

        try json_write_stream.objectField("upvoted");
        try json_write_stream.emitBool(upvoted);

        try json_write_stream.endObject();
    }

    try json_write_stream.endArray();
    response.close = true;
}

pub fn exportQuestions(ctx: *Context, response: *http.Response, request: http.Request) !void {
    const allocator = request.arena;

    var store = Store{ .redis_client = &ctx.redis_client };
    var session = store.get(allocator, request, "nochfragen_session");
    const logged_in = (try session.get(bool, "authenticated")) orelse false;
    if (!logged_in) return forbidden(response, "Forbidden");

    var iter = try QuestionIterator.init(allocator, &ctx.db, true);
    defer iter.deinit();

    try response.headers.put(
        "Content-Disposition",
        "attachment; filename=\"questions.csv\"",
    );

    try response.writer().print(
        "text,upvotes,state,created_at,modified_at,answering_at,answered_at\n",
        .{},
    );

    while (try iter.next()) |question| {
        if (question.state == .deleted) continue;

        try std.json.encodeJsonString(question.text, .{}, response.writer());
        try response.writer().print(",{},{s},", .{
            question.upvotes,
            question.state.toString(),
        });
        try printTime(question.created_at, response.writer());
        try response.writer().writeAll(",");
        try printTime(question.modified_at, response.writer());
        try response.writer().writeAll(",");
        try printTime(question.answering_at, response.writer());
        try response.writer().writeAll(",");
        try printTime(question.answered_at, response.writer());
        try response.writer().writeAll("\n");
    }
}

pub fn addQuestion(ctx: *Context, response: *http.Response, request: http.Request) !void {
    const allocator = request.arena;

    var token_stream = json.TokenStream.init(request.body());
    const request_data = json.parse(
        struct { text: []const u8 },
        &token_stream,
        .{ .allocator = allocator },
    ) catch return badRequest(response, "Invalid JSON");

    if (request_data.text.len == 0) return badRequest(response, "Empty question");
    if (request_data.text.len > max_question_len) return badRequest(response, "Maximum question length exceeded");

    try ctx.db.exec(insert_question_query, .{}, QuestionInternalWithoutId{
        .text = sqlite.Text{ .data = request_data.text },
        .created_at = std.time.timestamp(),
    });

    try ok(response);
}

pub fn modifyQuestion(ctx: *Context, response: *http.Response, request: http.Request, raw_id: []const u8) !void {
    const allocator = request.arena;

    const id = std.fmt.parseInt(u32, raw_id, 10) catch return badRequest(response, "Invalid ID");

    var store = Store{ .redis_client = &ctx.redis_client };
    var session = store.get(allocator, request, "nochfragen_session");
    if (session.is_new) return forbidden(response, "Access denied");

    var token_stream = json.TokenStream.init(request.body());
    const request_data = json.parse(
        struct {
            upvote: bool,
            state: u32,
        },
        &token_stream,
        .{ .allocator = allocator },
    ) catch return badRequest(response, "Invalid JSON");

    if (request_data.upvote) {
        const str_id = try std.fmt.allocPrint(allocator, "question:{}", .{id});

        const upvoted = (try session.get(bool, str_id)) orelse false;
        if (upvoted) return forbidden(response, "Already upvoted");

        try ctx.db.exec(increment_upvotes_query, .{}, .{ .id = id });

        try session.set(bool, str_id, true);
    } else {
        const logged_in = (try session.get(bool, "authenticated")) orelse false;
        if (!logged_in) return forbidden(response, "Forbidden");

        const state = std.meta.intToEnum(QuestionState, request_data.state) catch
            return badRequest(response, "Invalid state");
        switch (state) {
            // don't track these separately
            .hidden,
            .unanswered,
            .deleted,
            => try ctx.db.exec(modify_question_query, .{}, .{
                .state = request_data.state,
                .modified_at = std.time.timestamp(),
                .id = id,
            }),

            .answering => try ctx.db.exec(answering_question_query, .{}, .{
                .answering_at = std.time.timestamp(),
                .id = id,
            }),

            .answered => try ctx.db.exec(answered_question_query, .{}, .{
                .answered_at = std.time.timestamp(),
                .id = id,
            }),
        }
    }

    try ok(response);
}

pub fn deleteAllQuestions(ctx: *Context, response: *http.Response, request: http.Request) !void {
    const allocator = request.arena;

    var store = Store{ .redis_client = &ctx.redis_client };
    var session = store.get(allocator, request, "nochfragen_session");
    const logged_in = (try session.get(bool, "authenticated")) orelse false;
    if (!logged_in) return forbidden(response, "Forbidden");

    try ctx.db.exec(delete_all_questions_query, .{}, .{});

    try ok(response);
}

File: backend/src/questions.rs
use axum::Json;
use axum::extract::Path;
use axum::http::{HeaderMap, HeaderValue, header};
use axum::response::Response;
use axum::{extract::State, http::StatusCode, response::IntoResponse};
use diesel::prelude::*;
use diesel::{QueryDsl, RunQueryDsl, SelectableHelper};
use serde::{Deserialize, Serialize};
use time::UtcDateTime;
use time::format_description::well_known::iso8601::TimePrecision;
use time::format_description::well_known::{Iso8601, iso8601};
use tower_sessions::Session;

use crate::models::{NewQuestion, Question, QuestionState};
use crate::{AppResult, AppState};

const MAX_QUESTION_LEN: usize = 500;

#[derive(Serialize)]
pub struct QuestionResponse {
    id: i32,
    text: String,
    upvotes: i32,
    state: QuestionState,
    upvoted: bool,
}

pub async fn list_questions(
    State(app_state): State<AppState>,
    session: Session,
) -> AppResult<Json<Vec<QuestionResponse>>> {
    let mut connection = app_state.db_pool.get()?;

    let logged_in = session
        .get::<bool>("authenticated")
        .await
        .unwrap_or(None)
        .unwrap_or(false);

    use crate::schema::questions::dsl::*;
    let result = if logged_in {
        questions
            .select(Question::as_select())
            .load(&mut connection)
    } else {
        questions
            .filter(state.eq_any(vec![
                QuestionState::Unanswered,
                QuestionState::Answering,
                QuestionState::Answered,
            ]))
            .select(Question::as_select())
            .load(&mut connection)
    };

    let result: Vec<QuestionResponse> =
        futures::future::join_all(result?.into_iter().map(async |question| {
            let question_id = question.id;

            QuestionResponse {
                id: question.id,
                text: question.text,
                upvotes: question.upvotes,
                state: question.state,
                upvoted: session
                    .get::<bool>(&format!("question:{question_id}").to_string())
                    .await
                    .unwrap_or(None)
                    .unwrap_or(false),
            }
        }))
        .await;

    Ok(Json(result))
}

#[derive(Deserialize)]
pub struct AddQuestionRequest {
    text: String,
}

pub async fn add_question(
    State(app_state): State<AppState>,
    Json(request): Json<AddQuestionRequest>,
) -> AppResult<Response> {
    let mut connection = app_state.db_pool.get()?;

    let text = request.text;

    if text.len() == 0 {
        return Ok((StatusCode::BAD_REQUEST, "Empty question").into_response());
    }

    if text.len() > MAX_QUESTION_LEN {
        return Ok((StatusCode::BAD_REQUEST, "Maximum question length exceeded").into_response());
    }

    let current_time = UtcDateTime::now().unix_timestamp().try_into()?;

    use crate::schema::questions;
    diesel::insert_into(questions::table)
        .values(NewQuestion {
            text,
            state: QuestionState::Hidden,
            upvotes: 0,
            created_at: current_time,
            modified_at: -1,
            answering_at: -1,
            answered_at: -1,
        })
        .execute(&mut connection)?;

    Ok(StatusCode::OK.into_response())
}

pub async fn delete_all_questions(
    State(app_state): State<AppState>,
    session: Session,
) -> AppResult<StatusCode> {
    let mut connection = app_state.db_pool.get()?;

    let logged_in = session
        .get::<bool>("authenticated")
        .await
        .unwrap_or(None)
        .unwrap_or(false);
    if !logged_in {
        return Ok(StatusCode::FORBIDDEN);
    };

    use crate::schema::questions;
    diesel::delete(questions::table).execute(&mut connection)?;

    Ok(StatusCode::OK)
}

#[derive(Deserialize)]
pub struct ModifyQuestionRequest {
    upvote: bool,
    state: QuestionState,
}

pub async fn modify_question(
    Path(question_id): Path<i32>,
    State(app_state): State<AppState>,
    session: Session,
    Json(request): Json<ModifyQuestionRequest>,
) -> AppResult<Response> {
    let mut connection = app_state.db_pool.get()?;

    if request.upvote {
        let str_id = format!("question:{question_id}").to_string();

        let upvoted = session
            .get::<bool>(&str_id)
            .await
            .unwrap_or(None)
            .unwrap_or(false);
        if upvoted {
            return Ok((StatusCode::FORBIDDEN, "Already upvoted").into_response());
        }

        use crate::schema::questions::dsl::*;
        diesel::update(questions.find(question_id))
            .set(upvotes.eq(upvotes + 1))
            .execute(&mut connection)?;

        session.insert(&str_id, true).await?;
    } else {
        let logged_in = session
            .get::<bool>("authenticated")
            .await
            .unwrap_or(None)
            .unwrap_or(false);
        if !logged_in {
            return Ok(StatusCode::FORBIDDEN.into_response());
        };

        let current_time: i32 = UtcDateTime::now().unix_timestamp().try_into()?;

        use crate::schema::questions::dsl::*;
        match request.state {
            QuestionState::Hidden | QuestionState::Unanswered | QuestionState::HiddenAnswered => {
                diesel::update(questions.find(question_id))
                    .set((state.eq(request.state), modified_at.eq(current_time)))
                    .execute(&mut connection)?;
            }

            QuestionState::Answering => {
                diesel::update(questions.find(question_id))
                    .set((state.eq(request.state), answering_at.eq(current_time)))
                    .execute(&mut connection)?;
            }

            QuestionState::Answered => {
                diesel::update(questions.find(question_id))
                    .set((state.eq(request.state), answered_at.eq(current_time)))
                    .execute(&mut connection)?;
            }
        }
    }

    Ok(StatusCode::OK.into_response())
}

pub async fn delete_question(
    Path(question_id): Path<i32>,
    State(app_state): State<AppState>,
    session: Session,
) -> AppResult<StatusCode> {
    let mut connection = app_state.db_pool.get()?;

    let logged_in = session
        .get::<bool>("authenticated")
        .await
        .unwrap_or(None)
        .unwrap_or(false);
    if !logged_in {
        return Ok(StatusCode::FORBIDDEN);
    };

    use crate::schema::questions::dsl::*;
    diesel::delete(questions.find(question_id)).execute(&mut connection)?;

    Ok(StatusCode::OK)
}

fn unix_timestamp_to_str(timestamp: i32) -> anyhow::Result<String> {
    const CONFIG: iso8601::Config =
        iso8601::Config::DEFAULT.set_time_precision(TimePrecision::Second {
            decimal_digits: None,
        });
    let format = Iso8601::<{ CONFIG.encode() }>;

    let timestamp = UtcDateTime::from_unix_timestamp(timestamp.into())?;
    Ok(timestamp.format(&format)?)
}

pub async fn export_questions(
    State(app_state): State<AppState>,
    session: Session,
) -> AppResult<Response> {
    let mut connection = app_state.db_pool.get()?;

    let logged_in = session
        .get::<bool>("authenticated")
        .await
        .unwrap_or(None)
        .unwrap_or(false);
    if !logged_in {
        return Ok(StatusCode::FORBIDDEN.into_response());
    };

    use crate::schema::questions::dsl::*;
    let result = questions
        .select(Question::as_select())
        .load(&mut connection)?;

    #[derive(Serialize)]
    struct Row {
        text: String,
        upvotes: i32,
        state: String,
        created_at: String,
        modified_at: String,
        answering_at: String,
        answered_at: String,
    }

    impl TryFrom<Question> for Row {
        type Error = anyhow::Error;

        fn try_from(question: Question) -> Result<Self, Self::Error> {
            Ok(Self {
                text: question.text,
                upvotes: question.upvotes,
                state: question.state.to_string(),
                created_at: unix_timestamp_to_str(question.created_at)?,
                modified_at: unix_timestamp_to_str(question.modified_at)?,
                answering_at: unix_timestamp_to_str(question.answering_at)?,
                answered_at: unix_timestamp_to_str(question.answered_at)?,
            })
        }
    }

    let rows: Vec<Row> = result
        .into_iter()
        .map(Row::try_from)
        .collect::<anyhow::Result<Vec<Row>>>()?;

    let mut writer = csv::Writer::from_writer(vec![]);
    for row in rows {
        writer.serialize(row)?;
    }

    let mut headers = HeaderMap::new();
    headers.append(
        header::CONTENT_DISPOSITION,
        HeaderValue::from_str("attachment; filename=\"questions.csv\"")
            .expect("Static header should always be valid"),
    );

    Ok((headers, String::from_utf8(writer.into_inner()?)?).into_response())
}
